File: coding/src/com/algos/core/li17_interval/revised/L252_MeetingRooms.java
package  com.algos.core.li17_interval.revised;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import com.algos.core.models.Interval;

public class L252_MeetingRooms {

	public static void main(String[] args) {
		L252_MeetingRooms obj = new L252_MeetingRooms();

		List<Interval> intervals = new ArrayList<Interval>();
		intervals.add(new Interval(1,2));
		intervals.add(new Interval(2,3));
		System.out.println(obj.canAttendMeetings(intervals));
	}

    // Problem: https://www.lintcode.com/problem/920/
    // Check if the intervals overlap.
    // Idea: Sort intervals based on start in asc order.
    // Compare the adjacent intervals and see if they overlap. Track the previous interval in previous variable.
	public boolean canAttendMeetings(List<Interval> intervals) {
        if(intervals.size() <= 1)
            return true;

        Collections.sort(intervals, (i1, i2)->i1.start - i2.start);

        Interval prev = intervals.get(0);
        for(int i=1; i<intervals.size(); i++){
            Interval curr = intervals.get(i);
            if(prev.end > curr.start)
                return false;

            prev = curr;
        }

        return true;
    }
}

File: dsa/concepts/17_intervals/python/l252_meeting_rooms.py
# Problem: https://www.lintcode.com/problem/920/
# Check if the intervals overlap.
# Idea: Sort intervals based on start in asc order.
# Compare the adjacent intervals and see if they overlap.
# Track the previous interval in previous variable.
def can_attend_meetings(intervals):
    if len(intervals) <= 1:
        return True

    # sort by start time
    intervals.sort(key=lambda x: x[0])

    prev_start, prev_end = intervals[0]
    for i in range(1, len(intervals)):
        curr_start, curr_end = intervals[i]
        if prev_end > curr_start:
            return False
        prev_start, prev_end = curr_start, curr_end

    return True


# ------------------ Test (Equivalent to Java main) ------------------

intervals = [(1, 2), (2, 3)]
print(can_attend_meetings(intervals))  # Expected: True
